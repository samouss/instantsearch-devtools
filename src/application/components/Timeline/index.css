.Timeline {
  width: 35vw;
  height: 100vh;
  overflow-y: scroll;
  border-right: 1px solid lightgray;
}

.EventList {
  list-style: none;
  margin: 0;
  padding: 0;
}

.EventListItem {
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  padding: 10px;
  border-bottom: 1px solid lightgray;
}

.EventListItem:hover,
.EventListItemActive {
  background-color: #F5F5FA;
}

.EventListItem:hover {
  cursor: pointer;
}

.EventListItem:focus {
  outline: none;
}

.EventListItemSentinel {
  justify-content: center;
  border-bottom: none;
}

.EventListItemSentinel:hover {
  background: none;
  cursor: default;
}

.EventListItemButton {

}

.EventListItemDate {
  font-weight: 600;
}
