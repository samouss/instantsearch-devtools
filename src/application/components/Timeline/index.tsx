import React, { SFC } from 'react';
import cx from 'classnames';
import { Event, Id } from '../../types';
import Type from '../Type';
import DateTime from '../DateTime';
import styles from './index.css';

export type Props = {
  events: Event[];
  onClickEventTimeline: (event: Event) => void;
  selectedEventId?: Id;
};

const Timeline: SFC<Props> = ({
  events,
  selectedEventId,
  onClickEventTimeline,
}) => (
  <div styleName="Timeline">
    <ol styleName="EventList">
      {events.map(event => (
        <li
          key={event.id}
          className={cx(
            styles.EventListItem,
            event.id === selectedEventId && styles.EventListItemActive,
          )}
          onClick={() => onClickEventTimeline(event)}
        >
          <Type name={event.type} />
          <span styleName="EventListItemDate">
            <DateTime value={event.time} format="mm:ss.SSS" />
          </span>
        </li>
      ))}
    </ol>
  </div>
);

export default Timeline;
