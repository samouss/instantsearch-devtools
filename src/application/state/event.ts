import { v4 } from 'uuid';
import {
  HookEvent,
  ChangeHookEvent,
  SearchHookEvent,
  ResultHookEvent,
} from '../../types';
import {
  Id,
  State,
  Event,
  ChangeEvent,
  SearchEvent,
  ResultEvent,
} from '../types';

const START = Date.now();

const createChangeEvent = (event: ChangeHookEvent): ChangeEvent => ({
  type: 'CHANGE',
  id: v4(),
  time: Date.now() - START,
  parameters: event.parameters,
});

const createSearchEvent = (event: SearchHookEvent): SearchEvent => ({
  type: 'SEARCH',
  id: v4(),
  time: Date.now() - START,
  parameters: event.parameters,
});

const createResultEvent = (event: ResultHookEvent): ResultEvent => ({
  type: 'RESULT',
  id: v4(),
  time: Date.now() - START,
  parameters: event.parameters,
  results: event.results,
});

export const createEventFromHookEvent = (event: HookEvent): Event => {
  switch (event.type) {
    case 'CHANGE': {
      return createChangeEvent(event);
    }

    case 'SEARCH': {
      return createSearchEvent(event);
    }

    case 'RESULT': {
      return createResultEvent(event);
    }

    default: {
      throw new Error(`Event "${event.type}" is not supported.`);
    }
  }
};

const reducer = (state: State, event: Event): State => {
  switch (event.type) {
    case 'CHANGE': {
      return {
        ...state,
        eventIds: [event.id].concat(state.eventIds),
        eventById: state.eventById.set(event.id, event),
      };
    }

    case 'SEARCH': {
      return {
        ...state,
        eventIds: [event.id].concat(state.eventIds),
        eventById: state.eventById.set(event.id, event),
      };
    }

    case 'RESULT': {
      return {
        ...state,
        eventIds: [event.id].concat(state.eventIds),
        eventById: state.eventById.set(event.id, event),
      };
    }

    default: {
      return state;
    }
  }
};

export const getEvents = (state: State): Event[] => {
  return state.eventIds.map(id => {
    return state.eventById.get(id) as Event;
  });
};

export const getSelectedEvent = (state: State, id: Id): Event => {
  return state.eventById.get(id) as Event;
};

export default reducer;
