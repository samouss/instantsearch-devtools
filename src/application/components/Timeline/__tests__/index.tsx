import React from 'react';
import { shallow } from 'enzyme';
import * as eventTestUtils from 'test/event';
import Timeline from '../index';

describe('<Timeline />', () => {
  const defaultProps = {
    events: [],
    onClickEventTimeline: () => {},
  };

  it('expect to render without events', () => {
    const props = {
      ...defaultProps,
    };

    const wrapper = shallow(<Timeline {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('expect to render with events', () => {
    const props = {
      ...defaultProps,
      events: [
        eventTestUtils.createFakeChangeEvent(),
        eventTestUtils.createFakeSearchEvent(),
        eventTestUtils.createFakeResultEvent(),
      ],
    };

    const wrapper = shallow(<Timeline {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('expect to render with a selected event', () => {
    const props = {
      ...defaultProps,
      events: [
        eventTestUtils.createFakeChangeEvent(),
        eventTestUtils.createFakeSearchEvent({
          id: 'SEARCH_ID',
        }),
        eventTestUtils.createFakeResultEvent(),
      ],
      selectedEventId: 'CHANGE_ID',
    };

    const wrapper = shallow(<Timeline {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('expect to render with differences', () => {
    const differences = [
      {
        attribute: 'query',
        raw: {
          previous: '',
          next: 'Apple',
        },
        view: {
          previous: '""',
          next: '"Apple"',
        },
      },
    ];

    const props = {
      ...defaultProps,
      events: [
        eventTestUtils.createFakeChangeEvent({
          differences,
        }),
        eventTestUtils.createFakeSearchEvent({
          id: 'SEARCH_ID',
        }),
        eventTestUtils.createFakeResultEvent(),
      ],
      selectedEventId: 'CHANGE_ID',
    };

    const wrapper = shallow(<Timeline {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('expect to ignore differences on `RESULT`', () => {
    const differences = [
      {
        attribute: 'query',
        raw: {
          previous: '',
          next: 'Apple',
        },
        view: {
          previous: '""',
          next: '"Apple"',
        },
      },
    ];

    const props = {
      ...defaultProps,
      events: [
        eventTestUtils.createFakeChangeEvent(),
        eventTestUtils.createFakeSearchEvent({
          id: 'SEARCH_ID',
        }),
        eventTestUtils.createFakeResultEvent({
          differences,
        }),
      ],
      selectedEventId: 'CHANGE_ID',
    };

    const wrapper = shallow(<Timeline {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('expect to call `onClickEventTimeline` on click', () => {
    const onClickEventTimeline = jest.fn();
    const event = eventTestUtils.createFakeSearchEvent({
      id: 'SEARCH_ID',
    });

    const props = {
      ...defaultProps,
      events: [
        eventTestUtils.createFakeChangeEvent(),
        event,
        eventTestUtils.createFakeResultEvent(),
      ],
      onClickEventTimeline,
    };

    const wrapper = shallow(<Timeline {...props} />);

    expect(onClickEventTimeline).not.toHaveBeenCalled();

    wrapper
      .find('li')
      .at(1)
      .simulate('click');

    expect(onClickEventTimeline).toHaveBeenCalledWith(event);
  });
});
