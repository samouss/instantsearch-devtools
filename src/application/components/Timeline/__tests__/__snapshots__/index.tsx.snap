// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timeline /> expect to render with a selected event 1`] = `
<div
  styleName="Timeline"
>
  <ol
    styleName="EventList"
  >
    <li
      className="EventListItem"
      key="CHANGE_000"
      onClick={[Function]}
    >
      <Type
        name="CHANGE"
      />
      <span
        styleName="EventListItemDate"
      >
        <DateTime
          format="mm:ss.SSS"
          value={1000}
        />
      </span>
    </li>
    <li
      className="EventListItem"
      key="SEARCH_ID"
      onClick={[Function]}
    >
      <Type
        name="SEARCH"
      />
      <span
        styleName="EventListItemDate"
      >
        <DateTime
          format="mm:ss.SSS"
          value={2000}
        />
      </span>
    </li>
    <li
      className="EventListItem"
      key="RESULT_000"
      onClick={[Function]}
    >
      <Type
        name="RESULT"
      />
      <span
        styleName="EventListItemDate"
      >
        <DateTime
          format="mm:ss.SSS"
          value={3000}
        />
      </span>
    </li>
  </ol>
</div>
`;

exports[`<Timeline /> expect to render with events 1`] = `
<div
  styleName="Timeline"
>
  <ol
    styleName="EventList"
  >
    <li
      className="EventListItem"
      key="CHANGE_000"
      onClick={[Function]}
    >
      <Type
        name="CHANGE"
      />
      <span
        styleName="EventListItemDate"
      >
        <DateTime
          format="mm:ss.SSS"
          value={1000}
        />
      </span>
    </li>
    <li
      className="EventListItem"
      key="SEARCH_000"
      onClick={[Function]}
    >
      <Type
        name="SEARCH"
      />
      <span
        styleName="EventListItemDate"
      >
        <DateTime
          format="mm:ss.SSS"
          value={2000}
        />
      </span>
    </li>
    <li
      className="EventListItem"
      key="RESULT_000"
      onClick={[Function]}
    >
      <Type
        name="RESULT"
      />
      <span
        styleName="EventListItemDate"
      >
        <DateTime
          format="mm:ss.SSS"
          value={3000}
        />
      </span>
    </li>
  </ol>
</div>
`;

exports[`<Timeline /> expect to render without events 1`] = `
<div
  styleName="Timeline"
>
  <ol
    styleName="EventList"
  />
</div>
`;
